### DISCUSSION TITLE
help with an example pls. Deepgram openai a tts on unreal engine

### DISCUSSION THREAD

**Discussion by luca3viso** [deepgram/community#328] [General help] [answered]
Hello everyone, I need your help to have a working example for this. Microphone speaks, language understanding,  transcribe the text and feed it to ChatGPT, which generates a response. I would like this response to be spoken with a relevant voice - in English for English, in Italian for Italian, and in German for German. Additionally, I want this to be importable into Unreal Engine as a plugin to make Metahuman or other characters speak. There are similar services like ConveAI or Inworld, but they are not easy to maneuver with language switching.

**Comment by nikolawhallon** [comment]
Can you explain a bit more about what you'd like to see with the language switching? Deepgram offers language detection in batch-mode, but not in streaming-mode, however if your game/app allows the user to select a language, you could use that to specify what language to use in the Speech-to-Text. Another option is, if you use push-to-talk, after releasing the push-to-talk button, a single batch request can be made which does use language detection. There are a couple such options to explore.

I have not worked in Unreal (I'm a Godot guy with some Unity experience), but we have this question+answer which seems to get Deepgram streaming working for Unreal: https://github.com/orgs/deepgram/discussions/325#discussioncomment-6783924

Finally, here's a Unity project which uses Deepgram streaming + microphone + cleverbot (instead of chatgpt) + Amazon Polly to make an interactive NPC - it's a bit outdated (e.g. no chatgpt), but maybe it's useful: https://github.com/deepgram-devs/DGverse - I'll note that Deepgram does not have Text-to-Speech, so that part would need another service - I like Amazon Polly and Elevenlabs myself, but there are plenty to try out.

**Reply by luca3viso** [reply]
many thanks for your reply, I should make a virtual avatar that gives information for a campsite in 3 languages. italian english and german. so i could make 3 buttons and interact in real time from there on. i use deepgrams like sst chatgpt or similar with data training and a tts.Push to talk is a great IDEA thanks! The other problem in my case is the cost. the system will talk practically nonstop for 10 hours a day for 6 months. You would need a TTS hundreds of times cheaper than elevenlab which objectively works great but also costs a lot. SOME suggestions? Ideally, I would like to find an offline tts that perhaps takes advantage of a pre-trained or cloned voice.

**Reply by nikolawhallon** [reply, marked as answer]
Some ideas for the TTS: the higher quality the TTS, the more expensive, in general, I imagine. Elevenlabs might be more expensive than some alternatives, but I love the quality. On the other end of the spectrum, espeak can work offline and is free (https://pypi.org/project/pyttsx3/) but the quality is terrible, gTTS is maybe a middle ground, the quality is ok, and it's free (https://gtts.readthedocs.io/en/latest/). I imagine there must be other services too (and maybe Deepgram will have TTS at some point).

Another option for TTS is, if questions/answers are often the same or similar, or if certain explanations are the same every time, you could store that audio somewhere and just play that audio when needed, instead of making a new TTS request every time with the same/similar text (I'm not sure if this is a possibility for some of your dialog lines, but it's an idea).
